walk.go

```diff
@@ -36,65 +36,68 @@ type walkObserver func(info walkObserverInfo) error
 
 // walk determine if in is a `map[string]interface{}` or a `Slice` and traverse it if so, otherwise will
 // treat it as a scalar and invoke the walk observer on the input value directly.
-func walk(in interface{}, expandPaths bool, wo walkObserver) error {
-	switch in.(type) {
-	case map[string]interface{}:
-		return walkMap(in.(map[string]interface{}), expandPaths, wo)
-	case []interface{}:
-		return walkSlice(in.([]interface{}), expandPaths, wo)
+func walk(inVal reflect.Value , expandPaths bool, wo walkObserver) error {
+	switch inVal.Kind()  {
+	case reflect.Map:
+		return walkMap(inVal, expandPaths, wo)
+	case reflect.Slice:
+		return walkSlice(inVal, expandPaths, wo)
 	default:
-		return walkInterface(in, expandPaths, wo)
+		return walkInterface(inVal, expandPaths, wo)
 	}
 }
 
 // walkmap[string]interface{} is a shorthand way to walk a tree with a map as the root.
-func walkMap(m map[string]interface{}, expandPaths bool, wo walkObserver) error {
-	return walkFullMap(m, m, llpath.Path{}, expandPaths, wo)
+func walkMap(mVal reflect.Value, expandPaths bool, wo walkObserver) error {
+	return walkFullMap(mVal, mVal, llpath.Path{}, expandPaths, wo)
 }
 
 // walkSlice walks the provided root slice.
-func walkSlice(s []interface{}, expandPaths bool, wo walkObserver) error {
-	return walkFullSlice(s, map[string]interface{}{}, llpath.Path{}, expandPaths, wo)
+func walkSlice(sVal reflect.Value, expandPaths bool, wo walkObserver) error {
+	return walkFullSlice(sVal, reflect.ValueOf(map[string]interface{}{}), llpath.Path{}, expandPaths, wo)
 }
 
-func walkInterface(s interface{}, expandPaths bool, wo walkObserver) error {
+func walkInterface(s reflect.Value, expandPaths bool, wo walkObserver) error {
 	return wo(walkObserverInfo{
 		value: s,
 		key:   llpath.PathComponent{},
-		root:  map[string]interface{}{},
+		rootVal:  reflect.ValueOf(map[string]interface{}{}),
 		path:  llpath.Path{},
 	})
 }
 
-func walkFull(o interface{}, root map[string]interface{}, path llpath.Path, expandPaths bool, wo walkObserver) (err error) {
+func walkFull(oVal, rootVal reflect.Value, path llpath.Path, expandPaths bool, wo walkObserver) (err error) {
+
+	// Unpack any wrapped interfaces
+	for oVal.Kind() == reflect.Interface {
+		oVal = reflect.ValueOf(oVal.Interface())
+	}
+
 	lastPathComponent := path.Last()
 	if lastPathComponent == nil {
 		// In the case of a slice we can have an empty path
-		if _, ok := o.([]interface{}); ok {
+		if oVal.Kind() == reflect.Interface {
```

Comment:
I'm a little confused here. Before you did check if it's a slice of interfaces, but now `reflect.Interface` is good enough?

If you 'know' the type you want to check against you can also do this:

```
	var tInterfaceArray = reflect.TypeOf([]interface{}(nil))
    var tMapStringInterface = reflect.TypeOf((map[string]interface{})(nil))
```

These values can be used like:

```
  var val reflect.Type = ...
  if val.Type() == tMapStringInterface {
    ...
  }
```

or 

```
  var val reflect.Type = ...
  if val.Type().ConvertibleTo(tMapStringInterface) {
    m := val.Convert(tMapStringInterface).Interface().(map[string]interface{})
    ...
  }
```

This later trick actually helps in converting aliases of `map[string]interface{}` into `map[string]interface{}`. Then you could proceed without reflection ;)

Reply:
Thanks for sharing the neat trick! I think that's nifty, but given that we have this reflection based implementation, I think it's nice that it handles arbitrary value types for maps. So, now we can traverse (and as a result give nicer test output) for `map[string]string`, `map[string]int` etc etc.

And, it lays the ground work for better testing `map[int]uint64` and other weirder maps.

WDYT?

Reply:
I forgot to answer about slices since I got maps in my head. I'm fine to switch slices back to `[]interface{}`, I only switched to reflection for consistency with map traversal.

Reply:
Also, about the original line, this is a flat out bug from a bulk replace, will push a real fix.

Reply:
I think the most current code works right and still works correctly with `map[string]string` or whatever arbitrary value type, which is a win, since it gives us richer validation.